# SQL Data Analysis

### DecodeLabs Data Analytics Internship| Batch 2026

**Tools:** Python, Pandas, SQLite, SQL, Jupyter Notebook

---

## Project Overview

This project focuses on using SQL to analyze an e-commerce order dataset and extract actionable business insights.

The analysis covers product performance, customer purchasing behavior, payment methods, referral sources, order statuses, and sales trends over time.

The project demonstrates fundamental SQL concepts including:

- SELECT
- WHERE
- ORDER BY
- GROUP BY
- COUNT()
- SUM()
- AVG()
- MIN()
- MAX()
- HAVING
- CASE WHEN

## Business Objective

The objective of this project is to use SQL queries to analyze e-commerce transaction data and identify patterns in:

1. Overall business performance
2. Product sales and revenue
3. Customer purchasing behavior
4. Payment method usage
5. Referral source performance
6. Order status distribution
7. Monthly and yearly sales performance
8. High-value and repeat customers

The analysis aims to transform raw transactional data into meaningful business insights that can support decision-making.

In [ ]:
import pandas as pd
import sqlite3

In [4]:
df = pd.read_excel("Dataset for Data Analytics.xlsx")

df.head()

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04


In [5]:
df.shape

(1200, 14)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1200 entries, 0 to 1199
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   OrderID          1200 non-null   object        
 1   Date             1200 non-null   datetime64[ns]
 2   CustomerID       1200 non-null   object        
 3   Product          1200 non-null   object        
 4   Quantity         1200 non-null   int64         
 5   UnitPrice        1200 non-null   float64       
 6   ShippingAddress  1200 non-null   object        
 7   PaymentMethod    1200 non-null   object        
 8   OrderStatus      1200 non-null   object        
 9   TrackingNumber   1200 non-null   object        
 10  ItemsInCart      1200 non-null   int64         
 11  CouponCode       891 non-null    object        
 12  ReferralSource   1200 non-null   object        
 13  TotalPrice       1200 non-null   float64       
dtypes: datetime64[ns](1), float64(2), int64(

In [7]:
df.isnull().sum()

OrderID              0
Date                 0
CustomerID           0
Product              0
Quantity             0
UnitPrice            0
ShippingAddress      0
PaymentMethod        0
OrderStatus          0
TrackingNumber       0
ItemsInCart          0
CouponCode         309
ReferralSource       0
TotalPrice           0
dtype: int64

In [59]:
df.describe()

,Date,Quantity,UnitPrice,ItemsInCart,TotalPrice
count,1200,1200.000000,1200.000000,1200.000000,1200.000000
mean,2024-03-22 16:58:48,2.945833,356.412750,5.485000,1053.968300
min,2023-01-01 00:00:00,1.000000,11.390000,1.000000,11.390000
25%,2023-08-03 18:00:00,2.000000,186.062500,4.000000,410.520000
50%,2024-03-23 00:00:00,3.000000,364.210000,5.000000,823.615000
75%,2024-11-08 12:00:00,4.000000,521.570000,7.000000,1578.475000
max,2025-06-30 00:00:00,5.000000,699.930000,10.000000,3456.400000
std,NaN,1.407557,197.177146,2.281983,819.856558


In [57]:
df.columns

Index(['OrderID', 'Date', 'CustomerID', 'Product', 'Quantity', 'UnitPrice',
       'ShippingAddress', 'PaymentMethod', 'OrderStatus', 'TrackingNumber',
       'ItemsInCart', 'CouponCode', 'ReferralSource', 'TotalPrice'],
      dtype='object')

### Initial Data Inspection

The dataset contains 1,200 transaction records across 14 columns. The variables include order identifiers, dates, customer information, product details, pricing, payment methods, order status, referral sources, and total order values.

The dataset provides sufficient information for analyzing sales performance, customer behavior, payment preferences, marketing sources, and order outcomes.

In [8]:
conn = sqlite3.connect("project3_sales.db")

In [9]:
df.to_sql("orders", conn, if_exists="replace", index=False)

1200

In [60]:
query = """
SELECT name
FROM sqlite_master
WHERE type = 'table';
"""

pd.read_sql_query(query, conn)

,name
0,orders


In [12]:
query = """
SELECT *
FROM orders;
"""

result = pd.read_sql_query(query, conn)

result.head()

,OrderID,Date,CustomerID,Product,Quantity,UnitPrice,ShippingAddress,PaymentMethod,OrderStatus,TrackingNumber,ItemsInCart,CouponCode,ReferralSource,TotalPrice
0,ORD200000,2023-01-04 00:00:00,C72649,Monitor,5,570.62,928 Main St,Debit Card,Shipped,TRK37947903,7,SAVE10,Instagram,2853.10
1,ORD200001,2024-08-23 00:00:00,C75739,Phone,2,151.35,823 Main St,Online,Shipped,TRK91186779,3,SAVE10,Referral,302.70
2,ORD200002,2024-02-27 00:00:00,C81728,Tablet,5,550.68,512 Main St,Credit Card,Cancelled,TRK42903982,8,FREESHIP,Email,2753.40
3,ORD200003,2023-10-15 00:00:00,C33540,Chair,1,273.19,275 Main St,Debit Card,Returned,TRK62788070,5,SAVE10,Facebook,273.19
4,ORD200004,2025-05-08 00:00:00,C81840,Printer,4,626.01,668 Main St,Online,Delivered,TRK29241424,8,SAVE10,Email,2504.04


In [13]:
query = """
SELECT OrderID, Product, Quantity, TotalPrice
FROM orders;
"""

result = pd.read_sql_query(query, conn)

result.head(10)

,OrderID,Product,Quantity,TotalPrice
0,ORD200000,Monitor,5,2853.10
1,ORD200001,Phone,2,302.70
2,ORD200002,Tablet,5,2753.40
3,ORD200003,Chair,1,273.19
4,ORD200004,Printer,4,2504.04
5,ORD200005,Phone,2,491.72
6,ORD200006,Laptop,1,664.42
7,ORD200007,Monitor,5,747.75
8,ORD200008,Phone,2,268.56
9,ORD200009,Desk,4,2037.52


In [14]:
query = """
SELECT OrderID, Product, Quantity, TotalPrice
FROM orders
WHERE TotalPrice > 2000;
"""

result = pd.read_sql_query(query, conn)

result.head(10)

,OrderID,Product,Quantity,TotalPrice
0,ORD200000,Monitor,5,2853.10
1,ORD200002,Tablet,5,2753.40
2,ORD200004,Printer,4,2504.04
3,ORD200009,Desk,4,2037.52
4,ORD200010,Tablet,5,3129.85
5,ORD200032,Tablet,5,2683.60
6,ORD200034,Chair,4,2306.08
7,ORD200046,Monitor,4,2200.56
8,ORD200059,Chair,4,2373.08
9,ORD200072,Tablet,4,2357.76


In [18]:
query = """
SELECT OrderID, Product, Quantity, TotalPrice
FROM orders
WHERE Quantity > 4;
"""

result = pd.read_sql_query(query, conn)

result.head(10)

,OrderID,Product,Quantity,TotalPrice
0,ORD200000,Monitor,5,2853.10
1,ORD200002,Tablet,5,2753.40
2,ORD200007,Monitor,5,747.75
3,ORD200010,Tablet,5,3129.85
4,ORD200019,Monitor,5,1120.00
5,ORD200020,Chair,5,312.55
6,ORD200026,Laptop,5,297.75
7,ORD200032,Tablet,5,2683.60
8,ORD200057,Printer,5,1571.35
9,ORD200065,Monitor,5,408.50


In [19]:
query = """
SELECT OrderID, Product, Quantity, TotalPrice
FROM orders
ORDER BY TotalPrice DESC;
"""

result = pd.read_sql_query(query, conn)

result.head(10)

,OrderID,Product,Quantity,TotalPrice
0,ORD200789,Tablet,5,3456.40
1,ORD201122,Monitor,5,3390.95
2,ORD200632,Laptop,5,3390.80
3,ORD200469,Chair,5,3384.90
4,ORD200328,Tablet,5,3370.20
5,ORD200107,Printer,5,3353.75
6,ORD200326,Laptop,5,3352.40
7,ORD201065,Printer,5,3334.00
8,ORD201031,Phone,5,3322.55
9,ORD200463,Laptop,5,3313.90


In [20]:
query = """
SELECT 
    Product,
    COUNT(*) AS Number_of_Orders
FROM orders
GROUP BY Product
ORDER BY Number_of_Orders DESC;
"""

result = pd.read_sql_query(query, conn)

result

,Product,Number_of_Orders
0,Printer,181
1,Tablet,179
2,Chair,178
3,Laptop,173
4,Desk,170
5,Monitor,163
6,Phone,156


#### Finding

**Printer** had the highest number of orders with **181**, while **Phone** had the lowest with **156**.

In [62]:
query = """
SELECT 
    Product,
    SUM(TotalPrice) AS Total_Order_Value
FROM orders
GROUP BY Product
ORDER BY Total_Order_Value DESC;
"""

result = pd.read_sql_query(query, conn)

result

,Product,Total_Order_Value
0,Chair,195620.11
1,Printer,195612.61
2,Laptop,192126.56
3,Tablet,186568.95
4,Monitor,175651.41
5,Desk,167459.93
6,Phone,151722.39


#### Finding

**Chair** generated the highest recorded order value at **195,620.11**, closely followed by Printer at **195,612.61**

In [25]:
query = """
SELECT 
    Product,
    AVG(TotalPrice) AS Average_Order_Value
FROM orders
GROUP BY Product
ORDER BY Average_Order_Value DESC;
"""

result = pd.read_sql_query(query, conn)

result

,Product,Average_Order_Value
0,Laptop,1110.558150
1,Chair,1098.989382
2,Printer,1080.732652
3,Monitor,1077.616012
4,Tablet,1042.284637
5,Desk,985.058412
6,Phone,972.579423


#### Finding

**Laptop** had the highest average order value at approximately **1,110.56**.

In [63]:
query = """
SELECT 
    Product,
    COUNT(*) AS Purchase_Frequency,
    SUM(TotalPrice) AS Total_Order_Value,
    AVG(TotalPrice) AS Avg_Transaction_Value
FROM orders
GROUP BY Product
ORDER BY Total_Order_Value DESC;
"""

result = pd.read_sql_query(query, conn)

result

,Product,Purchase_Frequency,Total_Order_Value,Avg_Transaction_Value
0,Chair,178,195620.11,1098.989382
1,Printer,181,195612.61,1080.732652
2,Laptop,173,192126.56,1110.558150
3,Tablet,179,186568.95,1042.284637
4,Monitor,163,175651.41,1077.616012
5,Desk,170,167459.93,985.058412
6,Phone,156,151722.39,972.579423


In [44]:
query = """
SELECT 
    CustomerID,
    COUNT(*) AS Number_of_Orders
FROM orders
GROUP BY CustomerID
HAVING COUNT(*) > 1
ORDER BY Number_of_Orders DESC;
"""

result = pd.read_sql_query(query, conn)

result.head(10)

,CustomerID,Number_of_Orders
0,C98474,2
1,C97593,2
2,C94569,2
3,C91155,2
4,C70659,2
5,C56969,2
6,C46651,2
7,C38840,2
8,C35852,2
9,C21191,2


In [66]:
query = """
SELECT 
    CustomerID,
    COUNT(*) AS Number_of_Orders,
    SUM(TotalPrice) AS Total_Order_Value,
    AVG(TotalPrice) AS Avg_Order_Value
FROM orders
GROUP BY CustomerID
ORDER BY Total_Order_Value DESC;
"""

result = pd.read_sql_query(query, conn)

result.head(10)

,CustomerID,Number_of_Orders,Total_Order_Value,Avg_Order_Value
0,C38840,2,5723.23,2861.615
1,C57276,1,3456.40,3456.400
2,C67260,1,3390.80,3390.800
3,C13877,1,3384.90,3384.900
4,C18404,1,3370.20,3370.200
5,C16775,1,3353.75,3353.750
6,C65986,1,3352.40,3352.400
7,C47778,1,3334.00,3334.000
8,C59183,1,3322.55,3322.550
9,C25276,1,3313.90,3313.900


In [47]:
query = """
SELECT 
    OrderID,
    Date,
    Product,
    Quantity,
    UnitPrice,
    TotalPrice,
    PaymentMethod,
    OrderStatus
FROM orders
WHERE CustomerID = 'C38840'
ORDER BY Date;
"""

result = pd.read_sql_query(query, conn)

result

,OrderID,Date,Product,Quantity,UnitPrice,TotalPrice,PaymentMethod,OrderStatus
0,ORD201122,2023-06-07 00:00:00,Monitor,5,678.19,3390.95,Online,Returned
1,ORD200266,2024-06-08 00:00:00,Chair,4,583.07,2332.28,Credit Card,Cancelled


#### Finding

**Customer C38840** was the only customer meeting both high value order and repeat customers conditions, with **2 orders and 5,723.23 in recorded order value**.

However, both orders were **Returned or Cancelled**, so this customer should not automatically be classified as a successful retained customer.

In [28]:
query = """
SELECT 
    PaymentMethod,
    COUNT(*) AS Number_of_Orders
FROM orders
GROUP BY PaymentMethod
ORDER BY Number_of_Orders DESC;
"""

result = pd.read_sql_query(query, conn)

result

,PaymentMethod,Number_of_Orders
0,Online,258
1,Cash,246
2,Credit Card,234
3,Debit Card,232
4,Gift Card,230


In [67]:
query = """
SELECT 
    PaymentMethod,
    COUNT(*) AS Number_of_Orders,
    SUM(TotalPrice) AS Total_Order_Value,
    AVG(TotalPrice) AS Avg_Order_Value
FROM orders
GROUP BY PaymentMethod
ORDER BY Total_Order_Value DESC;
"""

result = pd.read_sql_query(query, conn)

result

,PaymentMethod,Number_of_Orders,Total_Order_Value,Avg_Order_Value
0,Credit Card,234,263847.63,1127.553974
1,Online,258,262442.94,1017.220698
2,Cash,246,259786.29,1056.041829
3,Gift Card,230,246323.92,1070.973565
4,Debit Card,232,232361.18,1001.556810


In [30]:
query = """
SELECT 
    ReferralSource,
    COUNT(*) AS Number_of_Orders
FROM orders
GROUP BY ReferralSource
ORDER BY Number_of_Orders DESC;
"""

result = pd.read_sql_query(query, conn)

result

,ReferralSource,Number_of_Orders
0,Instagram,259
1,Email,250
2,Google,241
3,Facebook,228
4,Referral,222


### Finding

**Credit Card** generated the highest recorded order value at **263,847.63**, while **Online** payments had the highest order count at **258**.

In [68]:
query = """
SELECT 
    ReferralSource,
    COUNT(*) AS Number_of_Orders,
    SUM(TotalPrice) AS Total_Order_Value,
    AVG(TotalPrice) AS Avg_Order_Value
FROM orders
GROUP BY ReferralSource
ORDER BY Total_Order_Value DESC;
"""

result = pd.read_sql_query(query, conn)

result

,ReferralSource,Number_of_Orders,Total_Order_Value,Avg_Order_Value
0,Instagram,259,275285.45,1062.878185
1,Email,250,261808.55,1047.234200
2,Google,241,250441.48,1039.176266
3,Facebook,228,250410.90,1098.293421
4,Referral,222,226815.58,1021.691802


### Finding

**Instagram** generated the highest order volume (**259**) and highest recorded order value (**275,285.45**).

In [27]:
query = """
SELECT 
    OrderStatus,
    COUNT(*) AS Number_of_Orders
FROM orders
GROUP BY OrderStatus
ORDER BY Number_of_Orders DESC;
"""

result = pd.read_sql_query(query, conn)

result

,OrderStatus,Number_of_Orders
0,Cancelled,250
1,Returned,247
2,Pending,237
3,Shipped,235
4,Delivered,231


#### Finding

**Cancelled orders** were the largest status category with **250** orders, followed by **Returned with 247**.

An important analytical caution is that TotalPrice represents recorded order value; the dataset does not provide a separate field confirming realized revenue for each status.

In [55]:
query = """
SELECT 
    COUNT(*) AS Total_Orders,
    SUM(
        CASE 
            WHEN OrderStatus IN ('Cancelled', 'Returned') 
            THEN 1 
            ELSE 0 
        END
    ) AS Cancelled_Returned_Orders
FROM orders;
"""

result = pd.read_sql_query(query, conn)

result

,Total_Orders,Cancelled_Returned_Orders
0,1200,497


In [56]:
query = """
SELECT 
    ROUND(
        100.0 * SUM(
            CASE 
                WHEN OrderStatus IN ('Cancelled', 'Returned') 
                THEN 1 
                ELSE 0 
            END
        ) / COUNT(*), 
        2
    ) AS Cancelled_Returned_Percentage
FROM orders;
"""

result = pd.read_sql_query(query, conn)

result

,Cancelled_Returned_Percentage
0,41.42


### Interpretation

**41.42%** of recorded orders were either **cancelled or returned**. This represents a substantial proportion of the dataset and may indicate potential issues related to fulfillment, customer satisfaction, product expectations, or the purchasing process.

Further investigation would be required to identify the underlying causes.

In [32]:
query = """
SELECT 
    MIN(Date) AS First_Order_Date,
    MAX(Date) AS Last_Order_Date
FROM orders;
"""

result = pd.read_sql_query(query, conn)

result

,First_Order_Date,Last_Order_Date
0,2023-01-01 00:00:00,2025-06-30 00:00:00


In [69]:
query = """
SELECT 
    strftime('%Y', Date) AS Year,
    COUNT(*) AS Number_of_Orders,
    SUM(TotalPrice) AS Total_Order_Value,
    AVG(TotalPrice) AS Avg_Order_Value
FROM orders
GROUP BY Year
ORDER BY Year;
"""

result = pd.read_sql_query(query, conn)

result

,Year,Number_of_Orders,Total_Order_Value,Avg_Order_Value
0,2023,510,552643.24,1083.614196
1,2024,459,480235.87,1046.265512
2,2025,231,231882.85,1003.821861


#### Finding

**2023** was the strongest full year, generating **552,643.24** from **510 orders**.

2025 should be interpreted cautiously because the dataset only runs through June 30, 2025.

In [70]:
query = """
SELECT 
    strftime('%Y-%m', Date) AS Month,
    COUNT(*) AS Number_of_Orders,
    SUM(TotalPrice) AS Total_Order_Value,
    AVG(TotalPrice) AS Avg_Order_Value
FROM orders
GROUP BY Month
ORDER BY Month;
"""

result = pd.read_sql_query(query, conn)

result

,Month,Number_of_Orders,Total_Order_Value,Avg_Order_Value
0,2023-01,47,56685.75,1206.079787
1,2023-02,37,40117.66,1084.261081
2,2023-03,43,48609.37,1130.450465
3,2023-04,31,27751.71,895.216452
4,2023-05,49,63836.84,1302.792653
5,2023-06,45,49500.19,1100.004222
6,2023-07,44,42820.66,973.196818
7,2023-08,51,54352.14,1065.728235
8,2023-09,29,29526.67,1018.161034
9,2023-10,47,52607.85,1119.315957


#### Finding

The highest-revenue month was **June 2024**, with recorded order value of **68,068.54**.

In [72]:
query = """
SELECT 
    COUNT(*) AS Total_Orders,
    SUM(TotalPrice) AS Total_Order_Value,
    AVG(TotalPrice) AS Avg_Order_Value,
    MAX(TotalPrice) AS Highest_Order_Value,
    MIN(TotalPrice) AS Lowest_Order_Value
FROM orders;
"""

result = pd.read_sql_query(query, conn)

result

,Total_Orders,Total_Order_Value,Avg_Order_Value,Highest_Order_Value,Lowest_Order_Value
0,1200,1264761.96,1053.9683,3456.4,11.39


# Key Findings & Business Insights

### Overall Performance
- The dataset contains 1,200 orders with a total recorded order value of 1,264,761.96.
- The average order value was 1,053.97.
- The highest individual order was valued at 3,456.40.

### Product Performance
- Printer recorded the highest number of orders with 181 transactions.
- Chair generated the highest recorded order value at 195,620.11.
- Laptop had the highest average order value at 1,110.56.

### Payment Methods
- Online was the most frequently used payment method with 258 orders.
- Credit Card generated the highest recorded order value at 263,847.63.
- Credit Card also had the highest average order value at approximately 1,127.55.

### Marketing Channels
- Instagram generated the highest number of orders and the highest recorded order value.
- Facebook had the highest average order value among referral sources at approximately 1,098.29.

### Customer Behavior
- C38840 was the only repeat customer identified with more than 4,000 in recorded order value.
- However, both of this customer's orders were either returned or cancelled.

### Order Status
- Cancelled orders were the largest status category with 250 orders.
- 41.42% of all orders were either cancelled or returned.

### Time Trends
- 2023 was the strongest complete year in the dataset.
- June 2024 was the highest-revenue month.
- 2025 should be interpreted cautiously because the dataset only covers January through June.

# Recommendations

### 1. Investigate Cancellations and Returns
The 41.42% combined cancellation and return rate warrants further investigation. The business should examine reasons for cancellations and returns to identify recurring operational or customer-experience issues.

### 2. Strengthen High-Performing Marketing Channels
Instagram generated the highest order volume and recorded order value. The business could evaluate successful Instagram campaigns and consider increasing investment in strategies that are producing strong results.

### 3. Investigate High-Value Orders
Products and customer segments associated with high average order values should be studied further to understand what drives larger purchases.

### 4. Evaluate Payment Behavior
Credit Card transactions generated the highest recorded order value and average order value. The business could investigate whether customers using this payment method have distinct purchasing patterns.

### 5. Improve Customer Retention
Repeat customers should be analyzed further to understand purchasing frequency, product preferences, and order outcomes. Retention strategies could then be developed for customers with successful completed purchases.

### 6. Monitor Sales Trends
Monthly and yearly trends should be monitored regularly to identify periods of strong and weak performance and support better planning.

# Conclusion

This SQL analysis transformed raw e-commerce transaction data into actionable business insights.

Using SQLite and SQL within a Jupyter Notebook, the analysis demonstrated fundamental querying techniques including SELECT, WHERE, ORDER BY, GROUP BY, COUNT, SUM, AVG, MIN, MAX, HAVING, and CASE WHEN.

The analysis identified differences in product performance, payment behavior, marketing channels, customer purchasing patterns, order statuses, and sales trends.

One of the most significant findings was that 41.42% of orders were either cancelled or returned, highlighting an area that requires further investigation. At the same time, Instagram emerged as the strongest referral source by order volume and recorded order value, while Chair generated the highest recorded order value among products.

Overall, the project demonstrates how SQL can be used to move from raw transactional data to structured analysis and business decision-making.